### PR TITLE
MINOR: additional check to follower fetch handling 

### PR DESCRIPTION
**Pull request by edoardocomar**
add check to KafkaApis
add unit test specific to follower fetch
update tool

developed with @mimaison

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by edoardocomar** [comment]
@rajinisivaram @junrao can you please check? thx

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 1]
```diff
@@ -489,7 +489,9 @@ class KafkaApis(val requestChannel: RequestChannel,
     val authorizedRequestInfo = mutable.ArrayBuffer[(TopicPartition, FetchRequest.PartitionData)]()
 
     for ((topicPartition, partitionData) <- fetchRequest.fetchData.asScala) {
-      if (!authorize(request.session, Read, new Resource(Topic, topicPartition.topic)))
+      val topicResource = new Resource(Topic, topicPartition.topic)
+      if ((fetchRequest.isFromFollower() && !authorize(request.session, ClusterAction, topicResource)) ||
```
Comment (rajinisivaram) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: @edoardocomar Is there any benefit from adding `ClusterAction` ACL to topics rather than just use the existing `ClusterAction` for the cluster resource? Since we don't need fine-grained topic-level ACL for this and brokers already need `ClusterAction` access for cluster, wouldn't that work? 
```
if (fetchRequest.isFromFollower && !authorize(request.session, ClusterAction, Resource.ClusterResource))
```

**Review by rajinisivaram — COMMENTED** [review]
@edoardocomar Thanks for the PR, left one comment/question.

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 1 above)
Comment (edoardocomar) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: Thanks @rajinisivaram 
the ClusterAction on the Cluster resource essentially is reserved fro broker-to-broker requests,
this would actually simplify the changes.

**Comment by edoardocomar** [comment]
@rajinisivaram looks much better now - no tool change and no new authorization to be given to the broker-to-broker principal!

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 2]
```diff
@@ -488,7 +488,12 @@ class KafkaApis(val requestChannel: RequestChannel,
     val nonExistingTopicResponseData = mutable.ArrayBuffer[(TopicPartition, FetchResponse.PartitionData)]()
     val authorizedRequestInfo = mutable.ArrayBuffer[(TopicPartition, FetchRequest.PartitionData)]()
 
-    for ((topicPartition, partitionData) <- fetchRequest.fetchData.asScala) {
+    if (fetchRequest.isFromFollower() && !authorize(request.session, ClusterAction, Resource.ClusterResource))    
+      for (topicPartition <- fetchRequest.fetchData.asScala.keys) 
+        unauthorizedTopicResponseData += topicPartition -> new FetchResponse.PartitionData(Errors.CLUSTER_AUTHORIZATION_FAILED,
+          FetchResponse.INVALID_HIGHWATERMARK, FetchResponse.INVALID_LAST_STABLE_OFFSET,
+          FetchResponse.INVALID_LOG_START_OFFSET, null, MemoryRecords.EMPTY)
+    else for ((topicPartition, partitionData) <- fetchRequest.fetchData.asScala) {
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: Nit: it's a bit odd to have the `for` in the same line as the `else`. I'd just move it to the next line.

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala [hunk 3]
```diff
@@ -1290,7 +1314,7 @@ class AuthorizerIntegrationTest extends BaseRequestTest {
 
   private def addAndVerifyAcls(acls: Set[Acl], resource: Resource) = {
     servers.head.apis.authorizer.get.addAcls(acls, resource)
-    TestUtils.waitAndVerifyAcls(servers.head.apis.authorizer.get.getAcls(resource) ++ acls, servers.head.apis.authorizer.get, resource)
+    TestUtils.waitAndVerifyAcls(servers.head.apis.authorizer.get.getAcls(resource), servers.head.apis.authorizer.get, resource)
```
Comment (rajinisivaram) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: Why was this change required? It looks like we don't wait and verify the new `acls` any more?

Comment (edoardocomar) [reply core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: @rajinisivaram it looked unnecessary : 
`servers.head.apis.authorizer.get.getAcls(resource)` already returns the `acls` that have been added to the authorizer `servers.head.apis.authorizer.get.addAcls(acls, resource)`

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 2 above)
Comment (edoardocomar) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: ok

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala (hunk 3 above)
Comment (edoardocomar) [reply core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: was the intent to verify that addAcls does actually add them ?

Comment (rajinisivaram) [reply core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: I would have thought so. Without `++ acls`, `TestUtils.waitAndVerifyAcls` doesn't really wait or verify anything.

Comment (edoardocomar) [reply core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: Oh well, it looks the test is written expecting `addAcls` to be asynchronous. 
It doesn't seem to be in the SimpleAclAuthorizer but I have no problem reverting that line.

**Review by rajinisivaram — APPROVED** [review]
@edoardocomar @mimaison Thanks for the PR, LGTM. Will merge after PR builds complete.

**Comment by edoardocomar** [comment]
the test failure seems spurious and completely unrelated.

**Comment by edoardocomar** [comment]
@rajinisivaram should this be included in the upcoming 1.0.1 ? @ewencp

**Comment by rajinisivaram** [comment]
@edoardocomar Yes, it should be in 1.0.1.
Build failure is unrelated. Merging to trunk, 1.0, 0.11.0 and 0.10.2 branches.

**Comment by rajinisivaram** [comment]
@edoardocomar I have pushed the commit to trunk and 1.0. The code has changed quite a bit since 0.11.0. Would you like to submit a PR for 0.11.0 and 0.10.2 branches? Thank you!

**Comment by mimaison** [comment]
Thanks @rajinisivaram. We'll work on the backports first thing tomorrow morning.
